### PR TITLE
Rewrite stage 2

### PR DESCRIPTION
**Pull request by dshynkev**
Replaced defines with enums and generally rewrote the whole thing to be more C++-style.

The only remaining defines are baud rate boilerplate.

### CODE REVIEW

File: include/MCP2515.h [hunk 1]
```diff
@@ -4,71 +4,67 @@
...
-    SPI::Driver spi_device;
+    class MCP2515: public SPI::Driver {
+    public:
+        MCP2515(Board::DigitalPin cs_pin = Board::D10);
+        uint8_t begin(uint8_t can_speed);
+        uint8_t set_filter(uint8_t num, uint32_t data);
+        uint8_t set_mask(uint8_t num, uint32_t data);
+        uint8_t send_buffer(uint32_t id, uint8_t len, uint8_t *buf);
+        void read_buffer(uint8_t len, uint8_t *buf);
+        uint8_t get_error(void);
```
Comment (Mogball) [inline include/MCP2515.h]: if there are no arguments no need for `(void)`

File: include/MCP2515.h [hunk 2]
```diff
@@ -4,71 +4,67 @@
...
-    void read_id(const uint8_t mcp_addr,
-                 uint8_t *ext,
-                 uint32_t *id);
-    void write_CAN_msg(const uint8_t buffer_sidh_addr);
-    void read_CAN_msg(const uint8_t buffer_sidh_addr);
-    void start_transmit(const uint8_t mcp_addr);
-    uint8_t get_next_free_buf(uint8_t *txbuf);
+        // Register commands
+        uint8_t read_register(const uint8_t address);
+        void read_registers(const uint8_t address,
```
Comment (Mogball) [inline include/MCP2515.h]: if you're gonna break the argument line put the first one down too:
```c++
void read_registers(
    const uint8_t address,
    uint8_t values[],
    const uint8_t n);
```

File: src/MCP2515.cpp [hunk 3]
```diff
@@ -466,98 +442,99 @@ void MCP2515::read_CAN_msg(const uint8_t buffer_sidh_addr)
...
-        ui_timeout++;
-    } while (res == MCP_ALLTXBUSY && (ui_timeout < TIMEOUTVALUE));
+        res = this->get_next_free_buf(&txbuf_n);
+        ++timeout;
+    } while (res == Result::AllBuffersBusy &&
+             (timeout < Limit::AwaitBufferTimeout));
 
-    if(ui_timeout == TIMEOUTVALUE)
-        return CAN_GETTXBFTIMEOUT;
+    if(timeout >= Limit::AwaitBufferTimeout)
```
Comment (Mogball) [inline src/MCP2515.cpp]: add braces

File: src/MCP2515.cpp [hunk 4]
```diff
@@ -466,98 +442,99 @@ void MCP2515::read_CAN_msg(const uint8_t buffer_sidh_addr)
...
+    uint16_t timeout = 0;
 
     do {
-        res = get_next_free_buf(&txbuf_n);
-        ui_timeout++;
-    } while (res == MCP_ALLTXBUSY && (ui_timeout < TIMEOUTVALUE));
+        res = this->get_next_free_buf(&txbuf_n);
+        ++timeout;
+    } while (res == Result::AllBuffersBusy &&
+             (timeout < Limit::AwaitBufferTimeout));
```
Comment (Mogball) [inline src/MCP2515.cpp]: you don't need the extra brackets here

File: src/MCP2515.cpp [hunk 5]
```diff
@@ -91,54 +89,54 @@ uint8_t MCP2515::init_mask(uint8_t num, uint8_t ext, uint32_t data) {
...
-        return res;
+        return Result::OK;
     }
 
-    switch(num) {
+    switch(filter_number) {
         case 0:
-            write_id(MCP_RXF0SIDH, ext, data);
+            write_id(Register::RXF0SIDH, filter);
         break;
```
Comment (Mogball) [inline src/MCP2515.cpp]: indentation on the break should be the same as the previous line

File: src/MCP2515.cpp [hunk 6]
```diff
@@ -91,54 +89,54 @@ uint8_t MCP2515::init_mask(uint8_t num, uint8_t ext, uint32_t data) {
...
     if(res > 0) {
         #if DEBUG_MODE
         trace << "Switching to config mode failed\n";
         #endif
-        return res;
+        return Result::OK;
     }
 
-    switch(num) {
+    switch(filter_number) {
```
Comment (Mogball) [inline src/MCP2515.cpp]: any chance you can get rid oft his switch statement?

File: src/MCP2515.cpp [hunk 7]
```diff
@@ -1,87 +1,85 @@
...
         #endif
-        return CAN_FAIL;
+        return Result::Failed;
     }
 
-    if (num == 0)
-        write_id(MCP_RXM0SIDH, ext, data);
-    else if(num == 1)
-        write_id(MCP_RXM1SIDH, ext, data);
+    if (mask_number == 0)
```
Comment (Mogball) [inline src/MCP2515.cpp]: add braces, even
```c++
if (mask_number == 0)
{ write_id(Register::RXM0SIDH, id); }
```
is okay

File: src/MCP2515.cpp [hunk 8]
```diff
@@ -1,87 +1,85 @@
...
 
-    uint8_t res = MCP2515_OK;
+    uint8_t res = Result::OK;
     #if DEBUG_MODE
     trace << "Starting mask setup\n";
     #endif
-    res = set_control_mode(MODE_CONFIG);
+    res = set_control_mode(Mode::Config);
     if (res > 0) {
         #if DEBUG_MODE
```
Comment (Mogball) [inline src/MCP2515.cpp]: preprocessor lines should not have indentation

File: src/MCP2515.cpp [hunk 9]
```diff
@@ -1,87 +1,85 @@
...
         #endif
-        return CAN_FAILINIT;
+        return Result::Failed;
     }
     #if DEBUG_MODE
         trace << "Config mode entered successfully\n";
     #endif
 
-    if (configure_rate(bus_rate) != MCP2515_OK) {
+    if (this->configure_rate(bus_rate) != Result::OK) {
```
Comment (Mogball) [inline src/MCP2515.cpp]: do you need to have `this->`?

**Review by Mogball — CHANGES_REQUESTED** [review]
Good work! I went through and picked at some code style

**Comment by dshynkev** [comment]
@Mogball Sorry for the delay, PD has had me.

File: src/MCP2515.cpp [hunk 10]
```diff
@@ -45,107 +45,88 @@ uint8_t MCP2515::begin(const uint8_t bus_rate) {
...
+#if DEBUG_MODE
     trace << "Mask setup successfull\n";
-    #endif
+#endif
 
     return Result::OK;
 }
 
 uint8_t MCP2515::set_filter(uint8_t filter_number, uint32_t filter)
 {
```
Comment (dhillondeep) [inline src/MCP2515.cpp]: Some places you have braces under the function on a new line while for some you have on the same line. Just make it consistent choose 1. Most of our code bases uses braces on the same line

**Review by dhillondeep — CHANGES_REQUESTED** [review]

File: src/MCP2515.cpp (hunk 10 above)
Comment (dshynkev) [reply src/MCP2515.cpp]: Fixed

**Review by dhillondeep — APPROVED** [review]

**Review by Mogball — APPROVED** [review]

**Pull request merged**